snapcraft/internal/build_providers/_multipass/_multipass.py

```diff
@@ -0,0 +1,107 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright (C) 2018 Canonical Ltd
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import shlex
+
+from .. import errors
+from .._base_provider import BaseProvider
+from ._instance_info import InstanceInfo
+from ._multipass_command import MultipassCommand
+
+
+class Multipass(BaseProvider):
+    """A multipass provider for snapcraft execute its lifecycle."""
+
+    def __init__(self, *, project, echoer) -> None:
+        self._multipass_cmd = MultipassCommand()
+        super().__init__(project=project, echoer=echoer,
+                         executor=self._multipass_cmd.execute)
+
+    def __enter__(self):
+        self.create()
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.destroy()
+
+    def create(self):
+        """Create the multipass instance and setup the build environment."""
+        self.launch_instance(self._multipass_cmd.launch, image='16.04')
+        self.setup_snapcraft()
+
+    def destroy(self):
+        """Destroy the instance, trying to stop it first."""
+        try:
+            instance_info = self._get_instance_info()
+        except errors.ProviderInfoError as info_error:
+            self.echoer.warning(
```

Comment:
Is this not an error? We weren't able to do as requested, but to the caller, this will look like a success.

Reply:
Let's live chat about this one